File: workflow/activities/summarize_chunks.py
from temporalio import activity
from ml.llm import llm


@activity.defn
async def summarize_chunks(chunks: list[str]) -> list[str]:
    """
    Activity to generate summaries for each code chunk.
    
    Args:
        chunks: List of code chunks
        
    Returns:
        List of summaries, one for each chunk
    """
    summaries = []
    
    for i, chunk_text in enumerate(chunks):
        result = llm.invoke(
            [
                (
                    "system",
                    """You are a code summarization assistant. Analyze the provided code snippet and create a 
                    brief, informative summary. Focus on what this specific portion of code does and its purpose.
                    Keep it concise but meaningful.""",
                ),
                ("user", f"Please summarize this code chunk (part {i+1}):\n\n{chunk_text}"),
            ],
        )
        content = result.content
        summaries.append(content if isinstance(content, str) else str(content))
    
    return summaries

File: ts/workers/runner/src/activities/summarize-chunks.ts
import { openai } from '@ai-sdk/openai';
import { generateText } from 'ai';
import { getConfig } from '@fdsa/shared';

/**
 * Activity to generate summaries for each code chunk.
 * 
 * @param chunks - List of code chunks
 * @returns List of summaries, one for each chunk
 */
export async function summarizeChunks(chunks: string[]): Promise<string[]> {
  const config = getConfig();
  const summaries: string[] = [];

  for (let i = 0; i < chunks.length; i++) {
    const chunkText = chunks[i];
    
    const result = await generateText({
      model: openai(config.openai.model),
      prompt: `You are a code summarization assistant. Analyze the provided code snippet and create a 
brief, informative summary. Focus on what this specific portion of code does and its purpose.
Keep it concise but meaningful.

Please summarize this code chunk (part ${i + 1}):

${chunkText}`,
    });

    summaries.push(result.text);
  }

  return summaries;
}
